design-proposals/klip-63-pause-resume-persistent-queries.md

```diff
@@ -0,0 +1,66 @@
+# KLIP-63: PAUSE and RESUME for persistent queries 
+
+**Author**: Jim Hughes (jnh5y) | 
+**Release Target**: Unknown | 
+**Status**: In Discussion | 
+**Discussion**: 
+
+**tl;dr:** Adds the ability to pause the processing of a persistent query.
+           
+## Motivation and background
+
+Users may wish to pause a query for a number of reasons and then resume processing.  As examples, users may wish to modify data pipelines built from persistent queries or to manage computational resources.
+
+Presently to manage processing, they would have to terminate a query and recreate it.  This approach means that any state would have to be restored; the goal of this work is to add the capability to pause processing temporarily while allowing Kafka consumers and state stores to be retained.
+
+## What is in scope
+
+Pausing and resuming persistent queries and the necessary language modifications to support this are in scope.
+
+## What is not in scope
+
+We are not considering the ability to pause and resume transient queries.  Additionally, modifying the consumers' state is out of scope.  (This would allow for processing to skip records or reprocess data.)
+
+## Value/Return
+
+End users will be able to manage query processing once this feature is implemented.  For example, this will allow for better control while designing data pipelines.  Pausing an upstream query will allow users to update a downstream query or other processing and iterate through design without processing all the data in a stream.
```

Comment:
Assuming that users pause and update a query that introduces breaking changes to downstream consumers/queries -- will we be able to detect that? How are we going to handle such cases?

Reply:
I'd suggest that situation is tantamount to any time that a Kafka producer starts writing a new data format to a topic.  In general, I don't think there's much defense for that.

The place where there are guardrails are topics with schemas registered.  I don't think ksqlDB will let a persistent query be created which writes incorrect output to an existing topic (which has a topic managed).